### PR TITLE
Port to Windows CNG APIs. Add memory management APIs.

### PR DESCRIPTION
**Pull request by rich617**
These changes port libjwt to Windows CNG APIs. 

Changes are:

- Add jwt-wincrypt.c file, which contains the CNG implementation of the signing and verification algorithms. 
- Add memory management functions jwt_free_str, jwt_set_alloc, and jwt_get_alloc to solve runtime mismatch issues that can occur with Windows DLLs.
- Explicitly cast size_t to int where needed to avoid loss of precision warnings on 64 bit platforms.

For both signing and verification, there is a new key format, which is a path to a key or cert stored in the Windows certificate store. This format is the same as used in PowerShell cmdlets, and is:

`cert:\<StoreType>\<StoreName>\<CertificateThumbprint>`

for example:

`cert:\CurrentUser\My\DACB7AB8BA5A1FC4D5939ECD627FB7E2402477E5`

or

`cert:\LocalMachine\My\59985D4AC8052B2F062FC973B5DC8E5DA5EBBDAA'

Note that for signing operations (jwt_encode), the operating system user under which the library runs must have access to the private key of the indicated certificate in order for LibJWT to access it. 

For Windows crypto, I was able to maintain backward compatibility with the key format for signature verification, which is a PEM-encoded public key. In other words, jwt_decode will accept either a PEM-encoded public key or a Windows certificate store path to a certificate. However, I was not able to implement PEM-encoded private key for JWT signing because the CNG APIs required for that are not available (that I know of). 

I have tested all this locally and verified (using test code and using the jwt.io online decoder) that the Windows implementation of the library can produce valid, signed JWTs for HMAC, RSA, and ECC algorithm types, and is also able to verify JWTs signed with those algorithms.

### CODE REVIEW

File: libjwt/jwt.c [hunk 1]
```diff
@@ -15,6 +15,55 @@
 #include "base64.h"
 #include "config.h"
 
+static jwt_alloc_t pfn_alloc = malloc;
+static jwt_realloc_t pfn_realloc = realloc;
+static jwt_free_t pfn_free = free;
```
Comment (benmcollins) [inline libjwt/jwt.c]: I would much rather set these to `NULL`, and make the below functions do:

    void *jwt_malloc(size_t size)
    {
        if (pfn_malloc)
            return pfn_malloc(size);
        else
            return malloc(size);
    }

This way, we can pass `NULL` to the `jwt_set_alloc()` for ones where we want to use native. Also. I believe some implementations may have `malloc` and friends set as a macro, and we can't pass them as pointers like you have.

This also makes it easy to selectively override some allocators, or revert to native, easily.

File: libjwt/jwt.c [hunk 2]
```diff
@@ -15,6 +15,55 @@
...
+void *jcalloc(size_t nmemb, size_t size)
+{
+	size_t total_size;
+	void* ptr;
+	
+	total_size = nmemb * size;
+	if (!total_size)
+		return NULL;
+
+	ptr = pfn_alloc(total_size);
```
Comment (benmcollins) [inline libjwt/jwt.c]: Change this to call `jwt_malloc`

File: libjwt/jwt.c [hunk 3]
```diff
@@ -15,6 +15,55 @@
...
+	pfn_free(ptr);
+}
+
+char *jstrdup(const char *str)
+{
+	size_t len;
+	char *result;
+
+	len = strlen(str);
+	result = (char *)pfn_alloc(len + 1);
```
Comment (benmcollins) [inline libjwt/jwt.c]: Change this to call `jwt_malloc`

File: libjwt/jwt.c [hunk 4]
```diff
@@ -1063,9 +1118,37 @@ char *jwt_encode_str(jwt_t *jwt)
...
+			jfree(str);
 		str = NULL;
 	}
 
 	return str;
 }
+
+int jwt_set_alloc(jwt_alloc_t palloc, jwt_realloc_t prealloc, jwt_free_t pfree)
+{
+	if (!palloc || !prealloc || !pfree)
```
Comment (benmcollins) [inline libjwt/jwt.c]: Remove `NULL` checks so that we can do selective override, and/or revert to native allocation.

**Comment by rich617** [comment]
Thanks for the comments. They all seem very reasonable except for one,
which is that allocate/free functions should be settable individually. In
my opinion, they should be set either all at once or not at all. It really
is asking for trouble to set "alloc" but not "free". I'm not attached to
this idea though, and so if you feel really strongly about it, I can
implement so that each can be set individually as you asked for.

However, I do like your comment about using NULL as the default, and
allowing a set for that case. The reason I did it the way I did was that it
was similar to the implementation in Jansson--though they did #undef all
the functions first to avoid the macro problem you indicated. I like your
idea better, and will implement it.

Also, good idea Googling "jalloc", etc. I'll implement all the comments,
and commit another iteration.

One other thing is I noticed that in one place I used CryptStringToBinaryA
and in another I used CryptStringToBinary. They are the same as long as the
code is not compiled with "UNICODE", but I am going to make it explicit
that we are using the non-wide version of the conversion functions
everywhere.

     --Rich

File: include/jwt.h [hunk 5]
```diff
@@ -538,14 +542,22 @@ JWT_EXPORT int jwt_encode_fp(jwt_t *jwt, FILE *fp);
  * Fully encode a JWT object and return as a string.
  *
  * Similar to jwt_encode_fp() except that a string is returned. The string
- * must be freed by the caller.
+ * must be freed by the caller using jwt_free_str().
```
[the remark this answers is not in the document]
Comment (rich617) [reply include/jwt.h]: Done

File: libjwt/jwt.c (hunk 1 above)
Comment (rich617) [reply libjwt/jwt.c]: Done

File: libjwt/jwt-private.h [hunk 6]
```diff
@@ -18,6 +18,13 @@ struct jwt {
 	json_t *headers;
 };
 
+/* Memory allocators. */
+void *jalloc(size_t size);
+void jfree(void *ptr);
+char *jstrdup(const char *str);
+void *jcalloc(size_t nmemb, size_t size);
+void *jrealloc(void *ptr, size_t size);
+
```
[the remark this answers is not in the document]
Comment (rich617) [reply libjwt/jwt-private.h]: Done. Note that jwt_free was already taken, and so I used jwt_freemem for the free function.

File: libjwt/jwt.c (hunk 2 above)
Comment (rich617) [reply libjwt/jwt.c]: Done

File: libjwt/jwt.c (hunk 3 above)
Comment (rich617) [reply libjwt/jwt.c]: Done

File: libjwt/jwt.c (hunk 4 above)
Comment (rich617) [reply libjwt/jwt.c]: Done. I decided that the all-or-nothing check was unnecessarily complicated. I left it as you suggested where the caller can specify any combination of functions they want.

**Pull request merged**